### ISSUE TITLE
ScenarioState should be injected when stage is wrapped into spring proxy

### ISSUE THREAD

**Issue by corneliusweig** [TNG/JGiven#254]
We are using a JGiven setup with Spring where the stages themselves are spring beans. However, as soon as the stage is wrapped into a spring proxy, the scenario state injection does no longer work.

See the attached test case for details. You can verify that both test cases should be working by disabling the logging aspect.

[testcase.tar.gz](https://github.com/TNG/JGiven/files/602419/testcase.tar.gz)

**Comment by janschaefer** [comment]
I could fix your example by setting `proxyTargetClass=true` as follows:

     @EnableAspectJAutoProxy(proxyTargetClass=true)

**Comment by janschaefer** [comment]
It seems that this is a general limitation of Spring and not related to JGiven in this case.

**Comment by corneliusweig** [comment]
Well, so it seems that for Aspects setting `proxyTargetClass=true` does the trick.

However, as far as I know @Transactional is always implemented via CGlib proxies, unless there's an interface to implement. This means that new testcase for this situation in the updated test cases below will never work? At least, unless there is an interface for the stage?

[testcase.tar.gz](https://github.com/TNG/JGiven/files/662256/testcase.tar.gz)

**Comment by janschaefer** [comment]
No, I currently assume that this is a different problem. I will check that.

**Comment by janschaefer** [comment]
With the current SNAPSHOT version of JGIven your example works as expected. However, there where some major Spring changes. Could you please test against version 0.14.0-SNAPSHOT? You have to add the sonatype snapshot repository to you Gradle/Maven configuration for that. E.g. for Gradle:

```
maven {
    url "https://oss.sonatype.org/content/repositories/snapshots"
}
```

**Comment by janschaefer** [comment]
You can now also test with v0.14.0-RC1

**Comment by corneliusweig** [comment]
All test-cases are now working with JGiven 0.14.0-SNAPSHOT including the one with @Transactional. Great work!

There's only a small flaw: whilst updating my test-cases, I upgraded to spring boot version 1.4.1.RELEASE (from 1.3.3.RELEASE) and two things popped up:

1. the annotation @SpringApplicationConfiguration is now deprecated from spring, but the recommended @SpringBootTest together with SpringRuleScenarioTest seems to create two instances of the spring application context (-> you see two spring logos during startup)
2. the application context cannot be started (even with @SpringApplicationConfiguration) due to a NPE at com.tngtech.jgiven.impl.intercept.ByteBuddyMethodInterceptor.

Not sure if this needs another issue though...

Anyway, here is the updated set of testcases using the newest version of spring boot
[testcase.tar.gz](https://github.com/TNG/JGiven/files/674819/testcase.tar.gz)

**Comment by janschaefer** [comment]
Ok. Might be a new issue. I will check that. Thanks  for testing!

**Comment by janschaefer** [comment]
I could not reproduce your issue. I have recreated your example configuration in the Spring-Boot example project (https://github.com/TNG/JGiven/tree/master/example-projects/spring-boot). But it works (no duplicate execution, no NPE). Could you provide a completely self-contained example, including the Gradle/Maven files?

**Comment by janschaefer** [comment]
Ok. I can reproduce it with our example project. I will further investigate the issue.

**Comment by janschaefer** [comment]
The duplicate creation of the Spring context was due to the NPE. After the NPE was fixed it works as expected. The NPE was caused, because for some reason Spring called the `hashCode` method on the Stage object before the Interceptor was correctly configured.

**Comment by janschaefer** [comment]
The issue is now fixed in release `0.14.0-RC2`. Could you verify that it works now?

**Comment by ahus1** [comment]
@janschaefer - I'm afraid something is broken with RC2, as the jgiven-junit pom.xml doesn't contain any dependencies, see http://search.maven.org/#artifactdetails%7Ccom.tngtech.jgiven%7Cjgiven-junit%7C0.14.0-RC2%7Cjar compared to http://search.maven.org/#artifactdetails%7Ccom.tngtech.jgiven%7Cjgiven-junit%7C0.13.1%7Cjar

**Comment by janschaefer** [comment]
Yes, you are right. This doesn't look look :-(. Thank you for reporting!

**Comment by janschaefer** [comment]
Is fixed now with 0.14.0-RC3.

**Comment by ahus1** [comment]
Although I wasn't the one who opened the original case, I can confirm that 0.14.0-RC3 works for me (Spring 1.4.1, JUnit Rules based integration with Spring and JGiven)

**Comment by janschaefer** [comment]
@ahus1 Thanks for the verification! I would still wait for verification of @corneliusweig, because they seem to have a quite special Spring configuration.

**Comment by catchin** [comment]
@janschaefer thanks for implementing this. I can confirm that 0.14.0-RC3 works for the (perhaps special?) use case @corneliusweig and I have.

**Comment by janschaefer** [comment]
@catchin Thanks for confirming! I will then release this version as 0.14.0. By 'special' I only meant that you use AspectJ proxies. Nothing to worry about ;-)

**Comment by corneliusweig** [comment]
Hi @janschaefer, sorry for not replying earlier. Version 0.14.0-RC3 and above works like a charm!

**Comment by janschaefer** [comment]
No problem, @catchin already answered. I have released v0.14.0 now (which is identical to RC3)
